### ISSUE TITLE
PlaidML OpenCL slower than tensorflow CPU

### ISSUE THREAD

**Issue by KirkSuD** [plaidml/plaidml#1723]
PlaidML using OpenCL is always slower than tensorflow using CPU.
I already tried training larger model like ResNet50.
Related issue: https://github.com/plaidml/plaidml/issues/467

I think it is using GPU, but not enough. Is there something wrong, or the overhead of using PlaidML is just too large, so it cannot utilize GPU?
When training using PlaidML, shared GPU memory increased (0.1GB -> 3.5GB); GPU usage% increased (0-1% -> 5-10%); CPU usage depends on model, about 20%-60%.
When using tensorflow CPU, CPU usage is 100%, not using GPU.

I will try to give as much info as possible.

CPU: AMD radeon 4700U
GPU: AMD radeon graphics
RAM: 8G

PlaidML env:
Pipenv Python 3.8.8
plaidml-keras 0.7.0
h5py downgraded to 2.10.0 to avoid error
(issue: https://github.com/plaidml/plaidml/issues/1705)

Tensorflow env: py3.8.8, tensorflow newest

clinfo:
```
Number of platforms:                             1
  Platform Profile:                              FULL_PROFILE
  Platform Version:                              OpenCL 2.1 AMD-APP (3188.4)
  Platform Name:                                 AMD Accelerated Parallel Processing
  Platform Vendor:                               Advanced Micro Devices, Inc.
  Platform Extensions:                           cl_khr_icd cl_khr_d3d10_sharing cl_khr_d3d11_sharing cl_khr_dx9_media_sharing cl_amd_event_callback cl_amd_offline_devices


  Platform Name:                                 AMD Accelerated Parallel Processing
Number of devices:                               1
  Device Type:                                   CL_DEVICE_TYPE_GPU
  Vendor ID:                                     1002h
  Board name:                                    AMD Radeon(TM) Graphics
  Device Topology:                               PCI[ B#3, D#0, F#0 ]
  Max compute units:                             7
  Max work items dimensions:                     3
    Max work items[0]:                           1024
    Max work items[1]:                           1024
    Max work items[2]:                           1024
  Max work group size:                           256
  Preferred vector width char:                   4
  Preferred vector width short:                  2
  Preferred vector width int:                    1
  Preferred vector width long:                   1
  Preferred vector width float:                  1
  Preferred vector width double:                 1
  Native vector width char:                      4
  Native vector width short:                     2
  Native vector width int:                       1
  Native vector width long:                      1
  Native vector width float:                     1
  Native vector width double:                    1
  Max clock frequency:                           1600Mhz
  Address bits:                                  64
  Max memory allocation:                         2351667609
  Image support:                                 Yes
  Max number of images read arguments:           128
  Max number of images write arguments:          64
  Max image 2D width:                            16384
  Max image 2D height:                           16384
  Max image 3D width:                            2048
  Max image 3D height:                           2048
  Max image 3D depth:                            2048
  Max samplers within kernel:                    16
  Max size of kernel argument:                   1024
  Alignment (bits) of base address:              2048
  Minimum alignment (bytes) for any datatype:    128
  Single precision floating point capability
    Denorms:                                     No
    Quiet NaNs:                                  Yes
    Round to nearest even:                       Yes
    Round to zero:                               Yes
    Round to +ve and infinity:                   Yes
    IEEE754-2008 fused multiply-add:             Yes
  Cache type:                                    Read/Write
  Cache line size:                               64
  Cache size:                                    16384
  Global memory size:                            3303538688
  Constant buffer size:                          2351667609
  Max number of constant args:                   8
  Local memory type:                             Scratchpad
  Local memory size:                             32768
  Max pipe arguments:                            16
  Max pipe active reservations:                  16
  Max pipe packet size:                          2351667609
  Max global variable size:                      2116500736
  Max global variable preferred total size:      3303538688
  Max read/write image args:                     64
  Max on device events:                          1024
  Queue on device max size:                      8388608
  Max on device queues:                          1
  Queue on device preferred size:                262144
  SVM capabilities:
    Coarse grain buffer:                         Yes
    Fine grain buffer:                           Yes
    Fine grain system:                           No
    Atomics:                                     No
  Preferred platform atomic alignment:           0
  Preferred global atomic alignment:             0
  Preferred local atomic alignment:              0
  Kernel Preferred work group size multiple:     64
  Error correction support:                      0
  Unified memory for Host and Device:            1
  Profiling timer resolution:                    1
  Device endianess:                              Little
  Available:                                     Yes
  Compiler available:                            Yes
  Execution capabilities:
    Execute OpenCL kernels:                      Yes
    Execute native function:                     No
  Queue on Host properties:
    Out-of-Order:                                No
    Profiling :                                  Yes
  Queue on Device properties:
    Out-of-Order:                                Yes
    Profiling :                                  Yes
  Platform ID:                                   00007FFCD0B59490
  Name:                                          gfx902
  Vendor:                                        Advanced Micro Devices, Inc.
  Device OpenCL C version:                       OpenCL C 2.0
  Driver version:                                3188.4 (PAL,HSAIL)
  Profile:                                       FULL_PROFILE
  Version:                                       OpenCL 2.0 AMD-APP (3188.4)
  Extensions:                                    cl_khr_fp64 cl_amd_fp64 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_fp16 cl_khr_gl_sharing cl_khr_gl_depth_images cl_amd_device_attribute_query cl_amd_vec3 cl_amd_printf cl_amd_media_ops cl_amd_media_ops2 cl_amd_popcnt cl_khr_d3d10_sharing cl_khr_d3d11_sharing cl_khr_dx9_media_sharing cl_khr_image2d_from_buffer cl_khr_subgroups cl_khr_gl_event cl_khr_depth_images cl_khr_mipmap_image cl_khr_mipmap_image_writes cl_amd_liquid_flash cl_amd_copy_buffer_p2p cl_amd_planar_yuv
```

plaidml-setup:
```
Default Config Devices:
   llvm_cpu.0 : CPU (via LLVM)

Experimental Config Devices:
   llvm_cpu.0 : CPU (via LLVM)
   opencl_amd_gfx902.0 : Advanced Micro Devices, Inc. gfx902 (OpenCL)

Using experimental devices can cause poor performance, crashes, and other nastiness.

Enable experimental device support? (y,n)[n]:y

Multiple devices detected (You can override by setting PLAIDML_DEVICE_IDS).
Please choose a default device:

   1 : llvm_cpu.0
   2 : opencl_amd_gfx902.0

Default device? (1,2)[1]:2

Selected device:
    opencl_amd_gfx902.0

Almost done. Multiplying some matrices...
Tile code:
  function (B[X,Z], C[Z,Y]) -> (A) { A[x,y : X,Y] = +(B[x,z] * C[z,y]); }
Whew. That worked.

Save settings to C:\Users\Kirk\.plaidml? (y,n)[y]:
Success!
```

plaidbench inference result:
```
Network Name         Inference Latency         Time / FPS
-----------------------------------------------------------------------------------------
mobilenet            377.02 ms                 374.65 ms / 2.67 fps  (CPU)
mobilenet            17.98 ms                  7.39 ms / 135.39 fps  (GPU)
resnet50             55.68 ms                  41.56 ms / 24.06 fps  (GPU)
Correctness: PASS, ..., fail_ratio: 0.0
```

plaidbench training always failed, no matter CPU or GPU.

I tried training really small model & fine-tuning ResNet50, loss & acc always fine but training slow.

Any idea? If this is just normal, then I have to give up training model using AMD GPU on my laptop...

**Comment by aliPMPAINT** [comment]
No idea about that, but if you wanted another way to train models via GPU, you can use ROCm as you have a VEGA card:
https://rocmdocs.amd.com/en/latest/

**Comment by KirkSuD** [comment]
OK. So we don't know why it's slower...
Unfortunately I can't use ROCm because of OS.
Looking forward to one day that we can save some time training using PlaidML on mid-tier AMD laptop.
